### PR TITLE
add a naive TimerHook LinkHook.

### PR DESCRIPTION
**Pull request by crcrpar**
This PR adds LinkHook based TimerHook mentioned in #5197.

This does not implement backward support mentioned as optional in the issue.

### CODE REVIEW

File: tests/chainer_tests/link_hooks_tests/test_timer.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        # summary
+        summary = hook.summary()
+        assert sorted(summary.keys()) == ['Linear', 'MyModel']
+        assert summary['Linear'] == {
+            'occurrence': 4,
+            'elapsed_time': times[0] + times[1] + times[3] + times[4]}
+        assert summary['MyModel'] == {
+            'occurrence': 2,
+            'elapsed_time': times[2] + times[5]}
+        summary['Linear']
```
Comment (toslunar) [inline tests/chainer_tests/link_hooks_tests/test_timer.py]: Could you remove this unused line?

File: tests/chainer_tests/link_hooks_tests/test_timer.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        # print_report
+        s = six.io.StringIO()
+        hook.print_report(file=s)
+        report = s.getvalue().split('\n')
```
Comment (toslunar) [inline tests/chainer_tests/link_hooks_tests/test_timer.py]: Why don't you use `s.getvalue().splitlines()`?

File: tests/chainer_tests/link_hooks_tests/test_timer.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        assert len(report) == 4
+        assert re.match(r' +Linear +[.0-9a-z]+ +4$', report[1])
+        assert re.match(r' +MyModel +[.0-9a-z]+ +2$', report[2])
```
Comment (toslunar) [inline tests/chainer_tests/link_hooks_tests/test_timer.py]: Why is the line (`report[2]`) required to start with spaces?

File: tests/chainer_tests/link_hooks_tests/test_timer.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        assert len(report[3]) == 0
```
Comment (toslunar) [inline tests/chainer_tests/link_hooks_tests/test_timer.py]:
```suggestion
        assert report[3] == ''
```

File: tests/chainer_tests/link_hooks_tests/test_timer.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        return self.l2(h)
+
+
+class TestTimerHook(unittest.TestCase):
+
+    def test_name(self):
+        assert link_hooks.TimerHook().name == 'TimerHook'
+
+    def check_forward(self, xp):
+        link = MyModel()
```
Comment (toslunar) [inline tests/chainer_tests/link_hooks_tests/test_timer.py]: It seems `to_gpu` (or `to_device`).

File: chainer/link_hooks/__init__.py [hunk 6]
```diff
@@ -0,0 +1 @@
+from chainer.link_hooks.timer import TimerHook  # NOQA
```
Comment (toslunar) [inline chainer/link_hooks/__init__.py]: Could you add the newline character at the end of the file?

File: tests/chainer_tests/link_hooks_tests/test_timer.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            'Linear', 'Linear', 'MyModel', 'Linear', 'Linear', 'MyModel']
+        assert times[0] + times[1] < times[2]
+        assert times[3] + times[4] < times[5]
+
+        # summary
+        summary = hook.summary()
+        assert sorted(summary.keys()) == ['Linear', 'MyModel']
+        assert summary['Linear'] == {
+            'occurrence': 4,
+            'elapsed_time': times[0] + times[1] + times[3] + times[4]}
```
Comment (toslunar) [inline tests/chainer_tests/link_hooks_tests/test_timer.py]: I'd rather not compare floats with `==`, because the addition is not really associative.

**Comment by crcrpar** [comment]
Do I need to add `'chainer.link_hooks'` to setup.py?

File: chainer/link_hooks/timer.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        """Returns total elapsed time in seconds."""
+        return self._total_time
+
+    def summary(self):
+        """Returns a summary of time profiling in links.
+        Returns:
+            A summarized dictionary whose keys are link names and
+            values are dictionaries of `elapsed_time` and `occurrence`.
+        """
+        summary = {}
```
Comment (toslunar) [inline chainer/link_hooks/timer.py]: How about using `collections.OrderedDict`?

File: tests/chainer_tests/link_hooks_tests/test_timer.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        assert names == [
+            'Linear', 'Linear', 'MyModel', 'Linear', 'Linear', 'MyModel']
+        assert times[0] + times[1] < times[2]
+        assert times[3] + times[4] < times[5]
+
+        # summary
+        summary = hook.summary()
+        assert sorted(summary.keys()) == ['Linear', 'MyModel']
+        assert summary['Linear']['occurrence'] == 4
+        assert math.isclose(
```
Comment (toslunar) [inline tests/chainer_tests/link_hooks_tests/test_timer.py]: Could you use `numpy.testing.assert_allclose` in order to support old python?

File: chainer/link_hooks/timer.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        assert link.xp == self.xp
+        self._postprocess(link)
+
+    def total_time(self):
+        """Returns total elapsed time in seconds."""
+        return self._total_time
+
+    def summary(self):
+        """Returns a summary of time profiling in links.
+        Returns:
```
Comment (toslunar) [inline chainer/link_hooks/timer.py]:
```suggestion

         Returns:
```

Could you fix the following error?
```
Warning, treated as error:
/home/travis/build/chainer/chainer/chainer/link_hooks/timer.py:docstring of chainer.link_hooks.TimerHook.summary:3:Unexpected indentation.
```

File: tests/chainer_tests/link_hooks_tests/test_timer.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+            summary['MyModel']['elapsed_time'],
+            times[2] + times[5])
+
+        # print_report
+        s = six.StringIO()
+        hook.print_report(file=s)
+        report = s.getvalue().splitlines()
+        assert len(report) == 3
+        assert re.match(r' +Linear +[.0-9a-z]+ +4$', report[1]) is not None
+        assert re.match(r' MyModel +[.0-9a-z]+ +2$', report[2]) is not None
```
Comment (toslunar) [inline tests/chainer_tests/link_hooks_tests/test_timer.py]: The order of the lines `report[1]` and `report[2]` cannot be asserted with old Python.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by crcrpar** [comment]
The failures on AppVeyor seems weird to me.

```
================================== FAILURES ===================================
_______________________ TestTimerHook.test_forward_cpu ________________________
self = <test_timer.TestTimerHook testMethod=test_forward_cpu>
    def test_forward_cpu(self):
>       self.check_forward(numpy)
tests\chainer_tests\link_hooks_tests\test_timer.py:74: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
self = <test_timer.TestTimerHook testMethod=test_forward_cpu>
xp = <module 'numpy' from 'C:\\Miniconda35-x64\\lib\\site-packages\\numpy\\__init__.py'>
    def check_forward(self, xp):
        link = MyModel()
        if xp is not numpy:
            link = link.to_gpu()
        hook = link_hooks.TimerHook()
    
        with hook:
            link(chainer.Variable(xp.array([[7, 5]], numpy.float32)))
            link(chainer.Variable(xp.array([[8, 1]], numpy.float32)))
    
        # call_history
        hist = hook.call_history
        assert len(hist) == 6
        assert all(len(h) == 2 for h in hist)
        names = [h[0] for h in hist]
        times = [h[1] for h in hist]
        assert names == [
            'Linear', 'Linear', 'MyModel', 'Linear', 'Linear', 'MyModel']
>       assert times[0] + times[1] < times[2]
E       AssertionError: assert (0.0 + 0.0) < 0.0
```

**Comment by crcrpar** [comment]
Should I use this snippet in `TimerHook` because time.time() is not accurate on Windows ?

https://github.com/chainer/chainer/blob/9ddb87cda9e47a58e140d0fd2e952f9cf33bb283/chainer/training/trainer.py#L16-L23

ref: https://docs.python.org/2.7/library/timeit.html?highlight=default_timer#timeit.default_timer

**Comment by toslunar** [comment]
> Should I use this snippet in `TimerHook` because time.time() is not accurate on Windows ?
> 
> [chainer/chainer/training/trainer.py](https://github.com/chainer/chainer/blob/9ddb87cda9e47a58e140d0fd2e952f9cf33bb283/chainer/training/trainer.py#L16-L23)

Looks good.  Could you fix `function_hooks.TimerHook` first in a new PR?

**Comment by toslunar** [comment]
Jenkins, test this please.

**Pull request merged**